vsock/src/vhu_vsock.rs

```diff
@@ -0,0 +1,488 @@
+        Self {
+            guest_cid,
+            socket,
+            uds_path,
+        }
+    }
+
+    /// Return the guest's current CID.
+    pub fn get_guest_cid(&self) -> u64 {
+        self.guest_cid
+    }
+
+    /// Return the path of the unix domain socket which is listening to
+    /// requests from the host side application.
+    pub fn get_uds_path(&self) -> String {
+        String::from(&self.uds_path)
+    }
+
+    /// Return the path of the unix domain socket which is listening to
+    /// requests from the guest.
+    pub fn get_socket_path(&self) -> String {
+        String::from(&self.socket)
+    }
+}
+
+impl TryFrom<VsockArgs> for VsockConfig {
+    type Error = Error;
+
+    fn try_from(cmd_args: VsockArgs) -> Result<Self> {
+        let socket = cmd_args.socket.trim().to_string();
+        let uds_path = cmd_args.uds_path.trim().to_string();
+
+        Ok(VsockConfig::new(cmd_args.guest_cid, socket, uds_path))
+    }
+}
+
+/// A local port and peer port pair used to retrieve
+/// the corresponding connection.
+#[derive(Hash, PartialEq, Eq, Debug, Clone)]
+pub struct ConnMapKey {
+    local_port: u32,
+    peer_port: u32,
+}
+
+impl ConnMapKey {
+    pub fn new(local_port: u32, peer_port: u32) -> Self {
+        Self {
+            local_port,
+            peer_port,
+        }
+    }
+}
+
+pub struct VhostUserVsockBackend {
+    guest_cid: __u64,
+    pub threads: Vec<Mutex<VhostUserVsockThread>>,
+    queues_per_thread: Vec<u64>,
+    pub exit_event: EventFd,
+}
+
+impl VhostUserVsockBackend {
+    pub(crate) fn new(vsock_config: VsockConfig) -> Result<Self> {
+        let mut threads = Vec::new();
+        let thread = Mutex::new(VhostUserVsockThread::new(
+            vsock_config.get_uds_path(),
+            vsock_config.get_guest_cid(),
+        )?);
+        let queues_per_thread = vec![QUEUE_MASK];
+        threads.push(thread);
+
+        Ok(Self {
+            guest_cid: vsock_config.get_guest_cid(),
+            threads,
+            queues_per_thread,
+            exit_event: EventFd::new(EFD_NONBLOCK).expect("Creating exit eventfd"),
+        })
+    }
+}
+
+impl VhostUserBackendMut<VringRwLock, ()> for VhostUserVsockBackend {
+    fn num_queues(&self) -> usize {
+        NUM_QUEUES
+    }
+
+    fn max_queue_size(&self) -> usize {
+        QUEUE_SIZE
+    }
+
+    fn features(&self) -> u64 {
+        1 << VIRTIO_F_VERSION_1
+            | 1 << VIRTIO_F_NOTIFY_ON_EMPTY
+            | 1 << VIRTIO_RING_F_EVENT_IDX
+            | VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits()
+    }
+
+    fn protocol_features(&self) -> VhostUserProtocolFeatures {
+        VhostUserProtocolFeatures::CONFIG
+    }
+
+    fn set_event_idx(&mut self, enabled: bool) {
+        for thread in self.threads.iter() {
+            thread.lock().unwrap().event_idx = enabled;
+        }
+    }
+
+    fn update_memory(
+        &mut self,
+        atomic_mem: GuestMemoryAtomic<GuestMemoryMmap>,
+    ) -> result::Result<(), io::Error> {
```

Comment:
Instead use std::io::Result directly.

    ) -> std::io::Result<()> {

For I2c/gpio, I have aliased that as IoResult (in a new pull request).

same applies to handle_event and elsewhere.

Reply:
Can we do this as a follow-up PR? (maybe opening an issue and when someone has time, we can fix)

Reply:
Fixed, it was just in the tests.